### PR TITLE
Throw if SocketInput is used after dispose

### PR DESCRIPTION
**Pull request by davidfowl**

### CODE REVIEW

File: src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/SocketInput.cs [hunk 1]
```diff
@@ -72,6 +74,8 @@ public void IncomingData(byte[] buffer, int offset, int count)
         {
             lock (_sync)
             {
+                CheckDisposed();
```
Comment (halter73) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/SocketInput.cs]: If we see any failures I would expect to see it here. While `IncomingStart/Complete` shouldn't ever be called after the socket is closed, there's no telling what a ConnectionFilter might do.

File: src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/SocketInput.cs [hunk 2]
```diff
@@ -316,15 +324,26 @@ public void Dispose()
...
                 }
-                _disposed = true;
+            }
+        }
+
+        private void CheckDisposed()
+        {
+            if (_disposed)
+            {
+                throw new ObjectDisposedException(GetType().Name);
```
Comment (halter73) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/SocketInput.cs]: `#if DEBUG`? While this shouldn't throw (but might in the case of ConnectionFilters), we probably shouldn't crash the entire server. If we should, we might as well put a `Trace.Assert` in the `MemoryPoolBlock` finalizer.

At least in the case of `ConnectionFilter`s, `IncommingData` will likely be called inside a `LoggingThreadPool` try/catch.

Comment (davidfowl) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/SocketInput.cs]: CRASH all the THINGS!

Comment (halter73) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/SocketInput.cs]: Just don't merge it lol

**Comment by davidfowl** [comment]
Doing this in debug mode seems fine

**Pull request closed without merging**